src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public class ServerSentEventsMessageParser
+    {
+        const byte ByteCR = (byte)'\r';
+        const byte ByteLF = (byte)'\n';
+        const byte ByteT = (byte)'T';
+        const byte ByteB = (byte)'B';
+        const byte ByteC = (byte)'C';
+        const byte ByteE = (byte)'E';
+
+        private static byte[] _dataPrefix = Encoding.UTF8.GetBytes("data: ");
+        private static byte[] _sseLineEnding = Encoding.UTF8.GetBytes("\r\n");
+        private static byte[] _newLine = Encoding.UTF8.GetBytes(Environment.NewLine);
+
+
+        private InternalParseState _internalParserState = InternalParseState.ReadMessageType;
+        private List<byte[]> _data = new List<byte[]>();
+
+        public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                var messageType = MessageType.Text;
+
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    // Partial message. We need to read more.
+                    return ParseResult.Incomplete;
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                if (line.Length <= 1)
+                {
+                    throw new FormatException("There was an error in the frame format");
+                }
+
+                if (IsMessageEnd(line))
+                {
+                    _internalParserState = InternalParseState.ReadEndOfMessage;
+                }
+                // To ensure that the \n was preceded by a \r
+                // since messages can't contain \n.
+                // data: foo\n\bar should be encoded as
+                // data: foo\r\n
+                // data: bar\r\n
+                else if (line[line.Length - _sseLineEnding.Length] != ByteCR)
+                {
+                    throw new FormatException("A '\\n' character can only be used as a line ending");
```

Comment:
`"Unexpected '\n' in message. A '\n' character can only be used as part of the newline sequence '\r\n'"` might be a little clearer.